File: Main/Decision Engine/Edges/edge_rules.py
from __future__ import annotations
from dataclasses import dataclass, asdict
from typing import Optional
import json

@dataclass
class EdgeColorScheme:
    stroke:        str
    stroke_active: str
    arrow_fill:    str
    label_fill:    str

@dataclass
class EdgeForce:
    distance: int
    strength: float

@dataclass
class EdgeWidth:
    default:     float
    highlighted: float
    dimmed:      float

EDGE_COLORS: dict[str, EdgeColorScheme] = {
    "critical": EdgeColorScheme(
        stroke="rgba(239, 68, 68, 0.75)",
        stroke_active="rgba(239, 68, 68, 1.0)",
        arrow_fill="#ef4444",
        label_fill="#f87171"
    ),
    "high": EdgeColorScheme(
        stroke="rgba(249, 115, 22, 0.75)",
        stroke_active="rgba(249, 115, 22, 1.0)",
        arrow_fill="#f97316",
        label_fill="#fb923c"
    ),
    "normal": EdgeColorScheme(
        stroke="rgba(51, 65, 85, 0.6)",
        stroke_active="rgba(100, 116, 139, 0.9)",
        arrow_fill="#334155",
        label_fill="#64748b"
    ),
}

EDGE_WIDTHS: dict[str, float] = {
    "critical":    2.0,
    "high":        1.5,
    "normal":      1.0,
    "highlighted": 2.5,
    "dimmed":      0.8,
}

EDGE_FORCES: dict[str, EdgeForce] = {
    "critical": EdgeForce(distance=90,  strength=0.85),
    "high":     EdgeForce(distance=110, strength=0.65),
    "normal":   EdgeForce(distance=145, strength=0.40),
}

EDGE_OPACITY: dict[str, float] = {
    "default":     0.7,
    "highlighted": 1.0,
    "dimmed":      0.1,
}
CRITICAL_RIGHTS: frozenset[str] = frozenset({
    "GenericAll",
    "WriteDacl",
    "WriteOwner",
    "DCSync",
    "AllExtendedRights",
    "Owns",
    "GenericWrite",
    "ForceChangePassword",
    "AddMember",
})

HIGH_RIGHTS: frozenset[str] = frozenset({
    "WriteProperty",
    "Self",
    "CreateChild",
    "DeleteChild",
    "ReadLAPSPassword",
    "GetChangesAll",
})

EDGE_LABEL_MAX_LENGTH: int = 20


def get_edge_category(edge_rights: list[str], crit_flag: bool = False) -> str:
    if crit_flag:
        return "critical"
    rights_set = set(edge_rights or [])
    if rights_set & CRITICAL_RIGHTS:
        return "critical"
    if rights_set & HIGH_RIGHTS:
        return "high"
    return "normal"


def get_edge_color(edge_rights: list[str], crit_flag: bool = False) -> EdgeColorScheme:
    return EDGE_COLORS[get_edge_category(edge_rights, crit_flag)]


def get_edge_width(edge_rights: list[str], crit_flag: bool = False) -> float:
    return EDGE_WIDTHS[get_edge_category(edge_rights, crit_flag)]


def get_edge_force(edge_rights: list[str], crit_flag: bool = False) -> EdgeForce:
    return EDGE_FORCES[get_edge_category(edge_rights, crit_flag)]


def truncate_label(label: str) -> str:
    if not label:
        return ""
    return label if len(label) <= EDGE_LABEL_MAX_LENGTH else label[: EDGE_LABEL_MAX_LENGTH - 1] + "…"


def build_rights_display(edge_rights: list[str]) -> str:
    if not edge_rights:
        return "ACE"
    if len(edge_rights) == 1:
        return edge_rights[0]
    if len(edge_rights) <= 3:
        return ", ".join(edge_rights)
    return f"{', '.join(edge_rights[:2])} +{len(edge_rights) - 2}"


def enrich_edge(edge: dict) -> dict:
    rights   = list(edge.get("edge_rights", []))
    crit     = bool(edge.get("crit", False))
    category = get_edge_category(rights, crit)
    color    = EDGE_COLORS[category]
    force    = EDGE_FORCES[category]

    edge.update({
        "crit":           category == "critical",
        "category":       category,
        "rights_display": build_rights_display(rights),
        "visual": {
            "stroke":      color.stroke,
            "strokeWidth": EDGE_WIDTHS[category],
            "arrowFill":   color.arrow_fill,
            "labelFill":   color.label_fill,
            "opacity":     EDGE_OPACITY["default"],
            "distance":    force.distance,
            "strength":    force.strength,
        }
    })
    return edge


def enrich_edges(edges: list[dict]) -> list[dict]:
    return [enrich_edge(e) for e in edges]

if __name__ == "__main__":
    sample_edges = [
        {
            "source": "s-1", "target": "s-2",
            "edge_rights": ["GenericAll"], "crit": False
        },
        {
            "source": "s-2", "target": "s-3",
            "edge_rights": ["WriteProperty", "ReadLAPSPassword"], "crit": False
        },
        {
            "source": "s-3", "target": "s-4",
            "edge_rights": ["ReadProperty"], "crit": False
        },
    ]
    enriched = enrich_edges(sample_edges)
    print(json.dumps(enriched, indent=2, ensure_ascii=False))

File: Main/Decision Engine/Edges/edge_rules.js
const EDGE_RULES = {

    colors: {
        blue: {
            stroke:       'rgba(56, 189, 248, 0.75)',
            strokeActive: 'rgba(56, 189, 248, 1.0)',
            arrowFill:    '#38bdf8',
            labelFill:    '#38bdf8'
        },
        critical: {
            stroke:       'rgba(239, 68, 68, 0.75)',   
            strokeActive: 'rgba(239, 68, 68, 1.0)',
            arrowFill:    '#ef4444',
            labelFill:    '#f87171'
        },
        high: {
            stroke:       'rgba(249, 115, 22, 0.75)', 
            strokeActive: 'rgba(249, 115, 22, 1.0)',
            arrowFill:    '#f97316',
            labelFill:    '#fb923c'
        },
        special: {
            stroke:       'rgba(234, 179, 8, 0.75)',
            strokeActive: 'rgba(234, 179, 8, 1.0)',
            arrowFill:    '#eab308',
            labelFill:    '#facc15'
        },
        normal: {
            stroke:       'rgba(51, 65, 85, 0.6)',     
            strokeActive: 'rgba(100, 116, 139, 0.9)',
            arrowFill:    '#334155',
            labelFill:    '#64748b'
        }
    },

    width: {
        blue:    3,
        critical: 3,     
        high:     3,
        special:  3,
        normal:   3,
        highlighted: 3,
        dimmed:      3
    },

    opacity: {
        default:     0.7,
        highlighted: 1.0,
        dimmed:      0.1
    },

    arrow: {
        viewBox:      '0 -5 10 10',
        path:         'M0,-5L10,0L0,5',
        refX:         38,          
        markerWidth:  6,
        markerHeight: 6,
        refXByType: {
            domain: 46,
            default: 38
        }
    },

    label: {
        fontSize:   '9px',
        fontFamily: 'JetBrains Mono, monospace',
        fontWeight: '400',
        opacity:    0.7,
        maxLength:  20,         
        showIfCritical: true,   
        showIfNormal: true      
    },

    force: {
        blue:     { distance: 200, strength: 0.40 },
        critical: { distance: 200, strength: 0.85 },
        high:     { distance: 200, strength: 0.65 },
        special:  { distance: 200, strength: 0.55 },
        normal:   { distance: 200, strength: 0.40 }
    },

    criticalRights: [
        'GenericAll',
        'WriteDACL',
        'WriteOwner',
        'All-Extended-Rights',
        'GenericWrite',
        'ForceChangePassword',
        'DS-Replication-Get-Changes',
        'DS-Replication-Get-Changes-All',
        'Read-gMSA-Password',
        'Write-msDS-KeyCredentialLink',
        'Write-msDS-AllowedToActOnBehalfOfOtherIdentity'
    ],

    highRights: [
        'WriteProperty',
        'Self',
        'CreateChild',
        'DeleteChild',
        'ReadLAPSPassword',
        'GetChangesAll',
        'Write-msDS-AllowedToDelegateTo',
        'Write-gPLink',
        'Write-gPOptions',
        'AddMember',
        'Self-Membership',
        'DS-Replication-Get-Changes-In-Filtered-Set'
    ],

    specialRights: [
        'WriteProperty',
        'Self',
        'Delete',
        'Validated-Write-SPN',
        'Send-As',
        'Receive-As',
        'Validated-DNS-Host-Name',
        'DS-Replication-Manage-Topology',
        'DS-Replication-Synchronize'
    ]
};

// UI overrides for specific extended-rights / object-type right names
// Ensure msDS-KeyCredentialLink is shown as "Shadow Credentials" in the graph
const EDGE_LABEL_OVERRIDES = {
    'write-msds-keycredentiallink': 'Shadow Credentials',
    'msds-keycredentiallink': 'Shadow Credentials',
    '5b47d60f-6051-40fb-99e0-ed3a78604e5d': 'Shadow Credentials'
};

function getEdgeCategory(d) {
    if (d.crit) return 'critical';
    const rights = Array.isArray(d.edge_rights) ? d.edge_rights : [];
    const rightsLower = rights.map(r => String(r).toLowerCase());
    if (rightsLower.some(r => r === 'kerberoastable' || r === 'asreproastable')) return 'blue';
    if (rights.some(r => EDGE_RULES.criticalRights.includes(r))) return 'critical';
    if (rights.some(r => EDGE_RULES.highRights.includes(r)))    return 'high';
    if (rights.some(r => EDGE_RULES.specialRights.includes(r)))  return 'special';
    return 'normal';
}

function getEdgeStroke(d) {
    return EDGE_RULES.colors[getEdgeCategory(d)].stroke;
}

function getEdgeWidth(d) {
    return EDGE_RULES.width[getEdgeCategory(d)];
}

function getEdgeMarker(d) {
    return `url(#arrow-${getEdgeCategory(d)})`;
}


function getEdgeLabelColor(d) {
    return EDGE_RULES.colors[getEdgeCategory(d)].labelFill || EDGE_RULES.colors[getEdgeCategory(d)].label_fill;
}

function getEdgeForce(d) {
    const force = EDGE_RULES.force[getEdgeCategory(d)];
    const sourceDegree = Number(d?.source?.edgeDegree || d?.source?.degree || 0);
    const targetDegree = Number(d?.target?.edgeDegree || d?.target?.degree || 0);
    const heavyNode = Math.max(sourceDegree, targetDegree) >= 10;

    if (!heavyNode) return force;

    return {
        distance: force.distance * 2,
        strength: force.strength
    };
}

function formatEdgeLabel(rel) {
    const max = EDGE_RULES.label.maxLength;
    if (!rel) return '';
    const key = String(rel).trim().toLowerCase();
    // Override key-credential related rights to a friendlier label
    if (key.includes('keycredential') || EDGE_LABEL_OVERRIDES[key]) {
        return EDGE_LABEL_OVERRIDES[key] || 'Shadow Credentials';
    }
    return rel.length > max ? rel.slice(0, max - 1) + '…' : rel;
}